Add windows' fullscreen states to ipc

Sem1rose opened the pull request:
Solves [#discussion:1843](https://github.com/YaLTeR/niri/discussions/1843).

Adds a `fullscreen_state` to the Window struct, and a `WindowFullscreenStateChanged` event.

Sem1rose committed:
Added Fullscreen state to IPC events

Sem1rose committed:
Fixed the halting problem

Sem1rose committed:
Added fullscreen state information for `niri msg`

nerdwave-nick commented on src/ipc/server.rs:
```diff
@@ -504,6 +504,9 @@ fn make_ipc_window(
 ) -> niri_ipc::Window {
+    // Quick hack to fix the consistent halting if we tried to access the sizing mode
+    // inside the constructor
+    let fullscreen_state = mapped.sizing_mode().into();
```
Without knowing as I'm new to niri and don't really know rust -- will the "quick hack" stay? Usually a hack is not really a good thing for a project.

Sem1Rose replied:
well, it's not a hack per se. the `mapped.sizing_mode().into();` was supposed to go inside the constructor below, but i believe `mapped` gets locked inside the `with_toplevel_role` method, so when i try and access it inside the same funciton a deadlock occurs. i could be totally wrong here, i'm not sure if this is intended of `with_toplevel_role`, but saving the `fullscreen_state` beforehand avoids any halts.

nerdwave-nick replied:
Hmmm... It's not necessary for any of the other values. That looks weird to me, but perhaps I'm just not knowledgeable enough regarding the code

nerdwave-nick commented on src/layout/mod.rs:
```diff
@@ -526,6 +526,27 @@ impl SizingMode {
+        }
+    }
+}
+
```
Instead of this, the SizingMode could instead be a unit only enum, which saves on these functions, and goes hand in hand with the idea I proposed to change it to sizing_mode instead of fullscreen_state

nerdwave-nick reviewed (commented):
Posting this with the disclaimer that I barely know rust and this is my first look ever at the niri code base, but I think the simplification I have proposed could be a general improvement for clarity and simplicity. Unless, of course, there are good reasons to do it the way it is. 

Thank you for being my experiment/first attempt at reading the niri code 😄

Sem1rose committed:
Merge branch 'main' into main

Sem1Rose commented on niri-ipc/src/lib.rs:
```diff
@@ -1532,6 +1538,16 @@ pub enum Event {
+        /// - `1`: maximized
+        /// - `2`: fullscreen
+        fullscreen_state: u8,
+    },
```
Small issue is that `SizingMode` is defined inside the `niri` crate and is therefore inaccessable inside `niri_ipc`. it's possible to move the definition and implementations from `niri` to `niri_ipc`, but this would be a breaking change and it's just stupid to refactor the entire project for a single featture.

nerdwave-nick replied:
Flexing my missing rust knowledge - could you serialzie the SizingMode as is? Who cares if it comes out as a lil string or number. But idk how rust is about this stuff

LucasOe commented:
> Does this also adds a match for window rules?

No. But I would be interested in adding support for `is-fullscreen` and `is-maximized` match rules based on this PR once it's merged.

Sem1rose committed:
Merge branch 'niri-wm:main' into main

Sem1rose committed:
window rule matches for the fullscreen-state

Sem1rose committed:
fixed halting problem when trying to access `sizing_mode` of the mapped window

side effect: we can't use the window's committed `ToplevelState`, so
we're using the the `server_pending` state, whose properties
consistently only update when focus changes.

Sem1rose committed:
added sizing_mode as a property of Mapped

renamed the `fullscreen_state` property and window rules match to
`sizing_mode` for consistency with existing `SizingMode` enum and
property.
removes the need to get the committed sizing mode, which causes halts if
used in the wrong place.
also, now the window rules are recalculated when the sizing_mode is
updated.

Sem1rose committed:
actually ran `cargo test` and fixed some bugs

YaLTeR commented:
The correct place to fix this is in niri and I will do it at some point, it's more complicated than it sounds

Sem1rose committed:
Merge branch 'niri-wm:main' into main
